### PR TITLE
Add more negative checks for integers passed to OPENSSL_malloc().

### PR DESCRIPTION
**Pull request by slontis**
##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: crypto/bio/bf_lbuf.c [hunk 1]
```diff
@@ -235,6 +235,8 @@ static long linebuffer_ctrl(BIO *b, int cmd, long num, void *ptr)
         obs = (int)num;
         p = ctx->obuf;
         if ((obs > DEFAULT_LINEBUFFER_SIZE) && (obs != ctx->obuf_size)) {
+            if (num < 0)
+                return 0;
             p = OPENSSL_malloc((int)num);
```
Comment (paulidale) [inline crypto/bio/bf_lbuf.c:240]: While you're here, change the cast to be (size_t).

File: crypto/bio/bf_buff.c [hunk 2]
```diff
@@ -318,6 +320,8 @@ static long buffer_ctrl(BIO *b, int cmd, long num, void *ptr)
         p1 = ctx->ibuf;
         p2 = ctx->obuf;
         if ((ibs > DEFAULT_BUFFER_SIZE) && (ibs != ctx->ibuf_size)) {
+            if (num < 0)
+                return 0;
             p1 = OPENSSL_malloc((int)num);
```
Comment (paulidale) [inline crypto/bio/bf_buff.c:325]: While you're here, change the cast to be (size_t).

File: crypto/bio/bf_buff.c [hunk 3]
```diff
@@ -289,6 +289,8 @@ static long buffer_ctrl(BIO *b, int cmd, long num, void *ptr)
         break;
     case BIO_C_SET_BUFF_READ_DATA:
         if (num > ctx->ibuf_size) {
+            if (num < 0)
+                return 0;
             p1 = OPENSSL_malloc((int)num);
```
Comment (paulidale) [inline crypto/bio/bf_buff.c:294]: While you're here, change the cast to be (size_t).

**Review by paulidale — APPROVED** [review]

File: crypto/asn1/bio_asn1.c [hunk 4]
```diff
@@ -116,7 +116,7 @@ static int asn1_bio_new(BIO *b)
 
 static int asn1_bio_init(BIO_ASN1_BUF_CTX *ctx, int size)
 {
-    if ((ctx->buf = OPENSSL_malloc(size)) == NULL) {
+    if (size < 0 || (ctx->buf = OPENSSL_malloc(size)) == NULL) {
```
Comment (t8m) [inline crypto/asn1/bio_asn1.c:119]: `<= 0` ?

File: crypto/asn1/tasn_utl.c [hunk 5]
```diff
@@ -168,6 +168,8 @@ int ossl_asn1_enc_save(ASN1_VALUE **pval, const unsigned char *in, int inlen,
         return 1;
 
     OPENSSL_free(enc->enc);
+    if (inlen < 0)
```
Comment (t8m) [inline crypto/asn1/tasn_utl.c:171]: `<= 0`

File: crypto/bio/bf_buff.c [hunk 6]
```diff
@@ -289,7 +289,9 @@ static long buffer_ctrl(BIO *b, int cmd, long num, void *ptr)
         break;
     case BIO_C_SET_BUFF_READ_DATA:
         if (num > ctx->ibuf_size) {
-            p1 = OPENSSL_malloc((int)num);
+            if (num < 0)
```
Comment (t8m) [inline crypto/bio/bf_buff.c:292]: `<= 0`

File: crypto/bio/bf_buff.c [hunk 7]
```diff
@@ -318,12 +320,14 @@ static long buffer_ctrl(BIO *b, int cmd, long num, void *ptr)
         p1 = ctx->ibuf;
         p2 = ctx->obuf;
         if ((ibs > DEFAULT_BUFFER_SIZE) && (ibs != ctx->ibuf_size)) {
-            p1 = OPENSSL_malloc((int)num);
+            if (num < 0)
```
Comment (t8m) [inline crypto/bio/bf_buff.c:323]: `<= 0`

File: crypto/bio/bf_lbuf.c [hunk 8]
```diff
@@ -235,7 +235,9 @@ static long linebuffer_ctrl(BIO *b, int cmd, long num, void *ptr)
         obs = (int)num;
         p = ctx->obuf;
         if ((obs > DEFAULT_LINEBUFFER_SIZE) && (obs != ctx->obuf_size)) {
-            p = OPENSSL_malloc((int)num);
+            if (num < 0)
```
Comment (t8m) [inline crypto/bio/bf_lbuf.c:238]: `<= 0`

File: crypto/x509/t_x509.c [hunk 9]
```diff
@@ -236,6 +236,8 @@ int X509_ocspid_print(BIO *bp, X509 *x)
         goto err;
     subj = X509_get_subject_name(x);
     derlen = i2d_X509_NAME(subj, NULL);
+    if (derlen < 0)
```
Comment (t8m) [inline crypto/x509/t_x509.c:239]: `<= 0`

File: crypto/x509/v3_conf.c [hunk 10]
```diff
@@ -154,6 +154,8 @@ static X509_EXTENSION *do_ext_i2d(const X509V3_EXT_METHOD *method,
         unsigned char *p;
 
         ext_len = method->i2d(ext_struc, NULL);
+        if (ext_len < 0)
```
Comment (t8m) [inline crypto/x509/v3_conf.c:157]: `<= 0`

File: crypto/x509/x509spki.c [hunk 11]
```diff
@@ -58,7 +58,10 @@ char *NETSCAPE_SPKI_b64_encode(NETSCAPE_SPKI *spki)
     unsigned char *der_spki, *p;
     char *b64_str;
     int der_len;
+
     der_len = i2d_NETSCAPE_SPKI(spki, NULL);
+    if (der_len < 0)
```
Comment (t8m) [inline crypto/x509/x509spki.c:63]: `<= 0`

**Review by t8m — COMMENTED** [review]
At some places where the length is directly passed to malloc() it is better to check for `<= 0` as the malloc will return NULL for 0 length and the malloc return value check will fail with misleading allocation error.

**Comment by slontis** [comment]
I was originally thinking that malloc(0) returned a non NULL pointer and I didnt want to break that case..
Unfortunately it is implementation dependant so if may return either NULL or non NULL - so the <= is probably safer..

**Comment by slontis** [comment]
Updated some < to <=

**Review by t8m — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Comment by slontis** [comment]
Ta. merged to master.

**Pull request closed without merging**